### DISCUSSION TITLE
Return type of this

### DISCUSSION THREAD

**Discussion by timia2109** [dotnet/csharplang#6804] [Language Ideas] [answered]
In C# especially with the builder pattern there are many methods which are simply returning `this`.

This have some drawbacks. For example if the class is inherited, the subtype gets lost. 

Example current version:
```csharp
namespace Example;

public class ABuilder {
    public ABuilder AddExample() {
        // ... Do stuff
        return this;
    }
}

public class BBuilder : ABuilder {
    // That return type is BBuilder
    public BBuilder AddOther() {
        // ... Do stuff
        return this;
    }

    // Return of .AddExample() is ABuilder, which could break the method 
    // chaining when changing the calling order
}

var builder = new BBuilder()
    .AddExample()
    .AddOther(); // <- Unknown
```

My idea is that C# should introduce the return type `this`. The Swift Language also have this construct [Self Type](https://docs.swift.org/swift-book/ReferenceManual/Types.html#grammar_self-type)

This could do two things:

 1. Tell the compiler that the return type is the same as the called object
 2. _(optional)_ Automatic add the statement `return this` at the end of the method

Example using the proposal:

```csharp
namespace Example;

public class ABuilder {
    public this AddExample() {
        // ... Do stuff
    }
}

public class BBuilder : ABuilder {
    // The compiler knows, that the return type is "BBuilder"
    public this AddOther() {
        // ... Do stuff
    }

    // Return of .AddExample() also is "BBuilder"
}

var builder = new BBuilder()
    .AddExample() // The compiler knows the return type of "BBuilder"
    .AddOther() // <- Works
```

Improvements for the Language:

 1. Compiler knows the exact type
 2. Less Writing: Implicit `return this` 
 3. Less Writing: The type `this` is shorter then `MyObjectBuilder`
 4. Conventional Building of Builders

**Comment by ufcpp** [comment, marked as answer]
https://github.com/dotnet/csharplang/issues/5413
